test/unit/uri.c

```diff
@@ -0,0 +1,209 @@
+	is(uri_ex->option_count, uri->option_count, "option count");
+	for (int i = 0; i < MIN(uri_ex->option_count, uri->option_count); i++)
+		uri_query_param_expected_check(&uri_ex->options[i], uri);
+	return check_plan();
+}
+
+static int
+test_string_uri_with_options_parse(void)
+{
+	const struct uri_expected uris[] = {
+		/* One string URI without options. */
+		/* [0] = */ {
+			/* string = */ "/path_to_unix_socket",
+			/* option_count = */ 0,
+			/* options = */ {},
+		},
+		/* One string URI without options. */
+		/* [1] = */ {
+			/* string = */ "/path_to_unix_socket",
+			/* option_count = */ 0,
+			/* options = */ {},
+		},
+		/* One string URI with one option and one option value. */
+		/* [2] = */ {
+			/* uri = */ "/path_to_unix_socket?q1=v1",
+			/* option_count = */ 1,
+			/* options = */ {
+				/* [0] = */ {
+					/* name = */ "q1",
+					/* value_count = */ 1,
+					/* values = */ { "v1" },
+				},
+			},
+		},
+		/**
+		 * Same as previous but with extra '&' at the end
+		 * of the string.
+		 */
+		/* [3] = */ {
+			/* string = */ "/path_to_unix_socket?q1=v1&",
+			/* option_count = */ 1,
+			/* options = */ {
+				/* [0] = */ {
+					/* name = */ "q1",
+					/* value_count = */ 1,
+					/* values = */ { "v1" },
+				},
+			},
+		},
+		/**
+		 * Same as previos but with two extra '&' at the end
+		 * of the string.
+		 */
+		/* [4] = */ {
+			/* string = */ "/path_to_unix_socket?q1=v1&&",
+			/* option_count = */ 1,
+			/* options = */ {
+				/* [0] = */ {
+					/* name = */ "q1",
+					/* value_count = */ 1,
+					/* values = */ { "v1" },
+				},
+			},
+		},
+		/*
+		 * One string URI with one option and two option values,
+		 * separated by "&".
+		 */
+		/* [5] = */ {
+			/* string = */ "/path_to_unix_socket?q1=v1&q1=v2",
+			/* option_count = */ 1,
+			/* options = */ {
+				/* [0] = */ {
+					/* name = */ "q1",
+					/* value_count = */ 2,
+					/* values = */ { "v1", "v2" },
+				},
+			},
+		},
+		/**
+		 * Same as previous but with extra '&' between options.
+		 */
+		/* [6] = */ {
+			/* string = */ "/path_to_unix_socket?q1=v1&&q1=v2",
+			/* option_count = */ 1,
+			/* options = */ {
+				/* [0] = */ {
+					/* name = */ "q1",
+					/* value_count = */ 2,
+					/* values = */ { "v1", "v2" },
+				},
+			},
+		},
+		/*
+		 * One string URI with several options.
+		 */
+		/* [7] = */ {
+			/* string = */ "/path_to_unix_socket"
+				       "?" "q1=v11&q1=v12&q2=v21&q2=v22",
+			/* option_count = */ 2,
+			/* options = */ {
+				/* [0] = */ {
+					/* name = */ "q1",
+					/* value_count = */ 2,
+				    	/* values = */ { "v11", "v12" },
+				},
+				/* [1] = */ {
+					/* name = */ "q2",
+					/* value_count = */ 2,
+					/* values = */ { "v21", "v22" },
+				},
+			},
+		},
+		/**
+		 * One string URI with several options, at the same time,
+		 * some of them have empty value or don't have values at all.
+		 */
+		/* [8] = */ {
+			/* uri = */ "/path_to_unix_socket"
+				    "?" "q1=v1&q1=&q2&q3=",
```

Comment:
A space between "?" and "q1..." is redundant - please join the strings.

Also, I don't see a test for the case when there's no value, nor equals sign (=), e.g. `foo&bar`. Is it supported? If so please add a test case.

Also, what happens if the query string is invalid, e.g. `===` or `&&&` or `=&=&'. Please add some test cases.